lgc/patch/PatchReadFirstLane.cpp

```diff
@@ -0,0 +1,477 @@
+    BuilderBase builder(curBb->getContext());
+    for (auto inst : m_insertLocations) {
+      // Avoid to insert reduncant readfirstlane
+      if (auto intrinsic = dyn_cast<IntrinsicInst>(inst))
+        if (intrinsic->getIntrinsicID() == Intrinsic::amdgcn_readfirstlane)
+          continue;
+      if (areAllUserReadFirstLane(inst))
+        continue;
+
+      applyReadFirstLane(inst, builder);
+      changed = true;
+    }
+  }
+  return changed;
+}
+
+// =====================================================================================================================
+// Collect the instructions that can be "assumed uniform" if the value itself isn't known to be uniform (by
+// DivergenceAnalysis), but we can replace all of its uses by readfirstlane(V) and still be correct
+//
+// @param block : The processing basic block
+// @param initialReadFirstLanes : The initial amdgcn_readfirstlane vector
+void PatchReadFirstLane::collectAssumeUniforms(BasicBlock *block,
+                                               const SmallVectorImpl<Instruction *> &initialReadFirstLanes) {
+  for (auto readfirstlane : initialReadFirstLanes)
+    m_canAssumeUniformDivergentUsesMap.insert({readfirstlane, {}});
+
+  for (Instruction &inst : llvm::reverse(*block)) {
+    // Skip readfirstlane-irrelevant instructions or a source of divergence which isn't clear to determine its operands
+    if (m_canAssumeUniformDivergentUsesMap.count(&inst) == 0 || (m_targetTransformInfo->isSourceOfDivergence(&inst)))
+      continue;
+
+    // For each operand, see if it can be assumed uniform to be inserted as the key of
+    // m_canAssumeUniformDivergentUsesMap. For all operands, see if they all can be readfirstlane-ified and then collect
+    // them as the value of m_canAssumeUniformDivergentUsesMap Readfirstlane-ified means that the operand is a key of
+    // the map and not fat pointer.
+    bool canCollectOperands = true;
+    auto loadInst = dyn_cast<LoadInst>(&inst);
+    SmallVector<Instruction *, 3> operandInsts;
+    for (Use &use : inst.operands()) {
+      if (!m_divergenceAnalysis->isDivergentUse(&use))
+        continue;
+      auto operandInst = dyn_cast<Instruction>(use.get());
+      if (!operandInst)
+        continue;
+      // Skip the operandInst that is already in the map
+      if (m_canAssumeUniformDivergentUsesMap.count(operandInst) == 1)
+        continue;
+      if (operandInst->getParent() != block || !isAllUsersDefinedInBlock(operandInst, block))
+        continue;
+
+      if (isAllUsersAssumedUniform(operandInst)) {
+        // The operand can be assumed uniform becasue all users are assumed uniform
+        m_canAssumeUniformDivergentUsesMap[operandInst] = {};
+        // NOTE: Stop propagation for a loadInst whose address space is fat pointer because it has uncertain size
+        if (loadInst && loadInst->getPointerAddressSpace() == ADDR_SPACE_BUFFER_FAT_POINTER)
+          canCollectOperands = false;
```

Comment:
Why is this check inside the loop? It refers to inst, not operandInst, which makes sense to me but suggests that the check should the outside of the loop.

Reply:
I think this check isn't really necessary. It will implicitly fall out from the fact that pointers aren't supported types.

Reply:
The reason I asked for the check against fat-pointer is for case like we cast an <2 x i32> to pointer type, and then the pointer was cast to fat pointer, then some value derived from that fat pointer was assume to-be uniform (I am not sure whether this happens or not), I thought we should not back-propagate through the fat pointer to the <2 x i32>, but after rethinking the problem I think we can propagate through the fat-pointer for such case to <2 x i32> if no other blocking condition, am I right? As long as we do not insert readfirstlane directly on fat pointers(which is excluded by the type-checking right before inserting readfirstlane), we would still be safe.

Reply:
Rebased code on Nicolai's two commits. I made a minor change to fix small bugs. 
-For candidate instruction itself, `call i32 @llvm.amdgcn.struct.tbuffer.load.i32(<4 x i32> %21, i32 %VertexIndex` should be a condition to stop propagate to its operands. Otherwise, the best location will be `%VertexIndex` which results in failures.
-In the middle loop of `findBestInsertLocation`, we should break out if the `current` cannot find in the map.